### ISSUE TITLE
Vertically splitted alignment in TTS training

### ISSUE THREAD

**Issue by ican24** [espnet/espnet#1937]
Hello,
I am trying to train my 20 hours dataset with train_pytorch_tacotron2.v3.yaml. The dataset is correct. It was used many times to synthesize a voice.

Since the 6-th epoch the alignment line was vertically splitted.   
![animation](https://user-images.githubusercontent.com/19214100/82045070-f9035b00-969d-11ea-9a02-77eef6e0f69e.gif)

Is it ok?
Is there a hope to receive positive result with this training?
The current epoch is 39.  

I may suppose that the main indicators of my training are normal  trying to learn many screenshots in espnet.

![Screenshot from 2020-05-15 11-05-05](https://user-images.githubusercontent.com/19214100/82044885-ae81de80-969d-11ea-8ca4-354bff856b28.png)

Thank you in advance

**Comment by unilight** [comment]
Have you tried to listen to the generated samples? The loss curve looks normal. Also, vertically split in the attention plot means that multiple output frames share the same hidden representation, i.e. input word. So it sounds okay too.

**Comment by ican24** [comment]
Thank you for a response!
I had stopped at 76 epoch trying  to launch decoding. 
Could you advise me a right method to listen and check samples? Do I need generate them?  
I do not see them in the "results" folder.
It's my first training in espnet.

**Comment by unilight** [comment]
You can follow the decoding and synthesis stages like here. https://github.com/espnet/espnet/blob/master/egs/ljspeech/tts1/run.sh#L199-L263

**Comment by ican24** [comment]
Thank you!
So simply I need to pass last stages after decoding without extra command line manipulation with Griffin-lim vocoder.

**Comment by ican24** [comment]
Dear @unilight 
Unfortunately, the generated samples are absolutely poor: no any sound.
Have you any supposition which is a real reason of such  deflection? 
The configuration file is unchanged, the dataset(22.05 kHz, ~22 hours) is a correct (I did many synhesis recenly). The single thing is non-english language: the "char" option was chosen instead of "phn".
As I told it is a my first raining in espnet, but I have enough experience to work with Kaldi style projects. So I'll try to fix he trouble, if you proide any hint.
 
Thank you in advance

**Comment by kan-bayashi** [comment]
If only the silence was generated, maybe you made something wrong.
How did you run the command?
Example command is as follows:
```
# you can specify the model to be used in decoding via `--model` option
./run.sh --stage 4 --model snapshot.ep.XXX --n_average 0
```
And please check the following items:
- Clean the input text as possible as you can
- Remove the silence at the beginning and the end of utterances
- Try a large reduction factor (e.g. 3 or 5) as a first step
- Continue to train more epochs

Related issues:
#1360 #1696

**Comment by ican24** [comment]
Hi,
My commad was without model option. I had tried with model, but the result is a same soudless.

The input text is without punctuation symbols and lowercase.
Have you an utility to remove the silence at the beginning and the end of utterances? 
I am trying to keep strongly your rules to avoid any misunderstanding by your side.
By the way, the aim of this training just to check you ParallelWaveGAN and see it in action.
I hope your project useful not only for common languages (English, Japanese, Mandarin) with their widely used datasets, but also for custom.

**Comment by kan-bayashi** [comment]
Hmm... That is strange.
Could you check the attention weight in decoding?
You can find the figures in outputs directory.

Please refer the following part to remove the silence (you need to tune the threshold).
https://github.com/espnet/espnet/blob/492367d6cf7da353e5ea71bf2674b6beafe39f2b/egs/m_ailabs/tts1/run.sh#L98-L107

> By the way, the aim of this training just to check you ParallelWaveGAN and see it in action.
I hope your project useful not only for common languages (English, Japanese, Mandarin) with their widely used datasets, but also for custom

PWG (and MelGAN) will work well for any languages even if the model is trained on the different languages.
And recently I made how to make recipe for PWG.
https://github.com/kan-bayashi/ParallelWaveGAN/blob/master/egs/README.md#how-to-make-the-recipe-for-your-own-dateset
This might help you.

**Comment by ican24** [comment]
Thank you!
I'll review all details and try to train it again.

**Comment by ican24** [comment]
Could you advise how I can use small piece of dataset for testing purpose correctly? As I remember there is a such advice in this forum to use just 3-5 records.
It will save a lot of time and make training/development much more productive.
Thank you in advance!

**Comment by kan-bayashi** [comment]
> Could you advise how I can use small piece of dataset for testing purpose correctly? 

You mean the adaptation with a small amount of training data?
What does your word “testing” mean?

**Comment by ican24** [comment]
Yes I mean training (step 3) with a small amount of training data?
Is it possible?

**Comment by kan-bayashi** [comment]
If you use adaptation with pretrained model, it can be.
There are some options for such an adaptation. See #1889 #1911.

**Comment by ican24** [comment]
Thank you!

**Comment by ican24** [comment]
Hi @kan-bayashi ,

My training with large reduction factor (3) failed again. Total epochs are 100.
The input text was highly cleaned.
The silence at the beginning and the end of utterances was removed
![Screenshot from 2020-05-23 21-39-01](https://user-images.githubusercontent.com/19214100/82804405-25775e00-9e7a-11ea-94c6-949979e9c5a3.png)

Here is an animated gif
![animation](https://user-images.githubusercontent.com/19214100/82804614-7d15c980-9e7a-11ea-992b-18fa7462657c.gif)

![Screenshot from 2020-05-25 11-50-33](https://user-images.githubusercontent.com/19214100/82806659-24e0c680-9e7e-11ea-92e3-1b7e34ab69da.png)

The wav file is so poor
[hayk.petrosyan_18036.wav.zip](https://github.com/espnet/espnet/files/4676770/hayk.petrosyan_18036.wav.zip)

Neverthless I intend to continue attempts to find a real reason of Tacotron2 to support non common languages  like my language.
Since December 2019 no one realization of Tacotron2 (Nvidia, Mozilla, Espnet any many less known) could not provide acceptable results . Meantime the aternative projects/networks do it quickly and without terrible headache  with the same datasets.  Right now I have 8 synthesized voices. Below is an example with 4 voices.
https://www.youtube.com/watch?v=AiHr3h0QZC4
    
I am going train with reduction factor = 5. 
Have you any additional recommendation before it? 

Thank you in advance!
I hope, as professional you are not going to ignore this issue with less common languages.

**Comment by unilight** [comment]
Hi, 
I listened to the sample. I cannot speak the language you are experimenting with, but it sounds like something (babble?). I think there is hope as long as it sounds like something close to speech! :)
You mentioned that you have used the same dataset to train TTS systems many times. Have you always been using characters as text input? Or are there any other forms of input of the language you use? For example, it would be impractical to use characters as text input for Mandarin TTS since there are thousands of characters in Mandarin. Rather, pinyin (a kind of phoneme representation of Mandarin) is commonly adopted.

**Comment by ican24** [comment]
Hi @unilight ,
Always I used characters. As G2P/pinyin I am  thinking time by time. Right now there are not good G2P tools for Armenian. Therefore I am not sure to use them. 
By the way I have my own phonetic conversion system for Kaldi ASR. Do you think it could be useful in this training? If yes, could you tell me any advice to do phonetic conversation with Kaldi like mechanism? May I simply convert using character conversation table?

**Comment by kan-bayashi** [comment]
> Meantime the aternative projects/networks do it quickly and without terrible headache with the same datasets.

What is the main difference compared to the projects? If there are already successful projects, you should follow the techniques used in the projects. Please raise the important difference or papers about your mentioned alternative system. 

> Do you think it could be useful in this training?

If you have the lexicon, it is better to convert from char to phoneme.

**Comment by kan-bayashi** [comment]
Another idea is the use of an adaptation technique (fine-tuning).
You can use the pretrained model of the other language to perform adaptation.
In this case, you can load the parameters of the pretrained model except for the embedding layer of the encoder, and then train the model with your language dataset.

**Comment by ican24** [comment]
Thank you for recommendations.
I am going to convert from char to phoneme and try it with same reduction factor = 3.
Do I need a less risk in reduction factor in "phn" training?

**Comment by kan-bayashi** [comment]
> Do I need a less risk in reduction factor in "phn" training?

3 is OK, I think. How about my question? I'm interested such a technique.
> What is the main difference compared to the projects? If there are already successful projects, you should follow the techniques used in the projects. Please raise the important difference or papers about your mentioned alternative system.

**Comment by ican24** [comment]
Recently I told you in our converstation in  
https://github.com/kan-bayashi/ParallelWaveGAN/issues/139

Mainly I mean https://github.com/tugstugi/pytorch-dc-tts
----------------------------------------------------
PyTorch implementation of Efficiently Trainable Text-to-Speech System Based on Deep Convolutional Networks with Guided Attention based partially on the following projects:
https://github.com/Kyubyong/dc_tts (audio pre processing)
https://github.com/r9y9/deepvoice3_pytorch (data loader sampler).
----------------------------------------------
It uses ljspeech like dataset (simple metadata.csv with UTF-8 support). Nothing extraordinary in data/audio preparation and processing. 
By the way, a week ago I succedded to feed Waveglow with its Text2Mel model, but right now my skills are not enough to tune it better.

**Comment by kan-bayashi** [comment]
Thanks. I did not remember you, sorry.
It is interesting to hear that DC-TTS is easier than Tacotron 2 in terms of training.

**Comment by ican24** [comment]
It is a fantastically quick and predictable!
I can train Text2Mel and SSRN models togather less than 24 hours using datasets not less 20 hours in my standard desktop machine with single RTX 2080 TI 12GB device.

**Comment by ican24** [comment]
Dear @kan-bayashi,

> If you have the lexicon, it is better to convert from char to phoneme.

Do I need to keep space between the letters in the phonetic conversion from lexion.txt? If yes, is there a need to double or change spaces between the words? 
Thank you in advance!

**Comment by ican24** [comment]
H,

I succeeded with phoneme conversion, where phones were separated by space. As regular spaces I doubled them. The text file looks like
`karine.hovhanissyan_05263 s i n y o r   s i s e rr i k y'   n y' s t e c ․
karine.hovhanissyan_05264 n a x a g a h y'   h a r y' c r e c ․
karine.hovhanissyan_05265 ―  d u q   a s u m   y e q ,  o r   x y' n d y' r a g i r y'   g y' r v a c'   e'   d y' r o sh y' m a t' y' gh t' i   v y' r a ՞։
karine.hovhanissyan_05266 ―  a y o ,  s i n y o r   n a x a g a h ,  ―  n o r i c   o t q i   t' y' rr ch e l o v   p a t a s y' x a n e c   s i n y o r   s i s e rr i k y' ,  ―  d y' r o sh y' m a t' y' gh t' i ։
karine.hovhanissyan_05267 ―  y e r k u   k o m y' s u h i n e r i   s y' t o r a g y' r u t' y u n y' n e r i   i s k u t' y u n y'   h a s y' t a t y' v a ՞c'   e' ։
karine.hovhanissyan_05268 ―  h a s y' t a t y' v a c'   e' ։`

By the way I used short dataset about 10 hours lenght, because the female voice is much more good for ML training in according to my experience. More over after normalization (more 1 second, less 10 seconds) the final legth is 6.01 hours.

![Screenshot from 2020-05-26 08-36-39](https://user-images.githubusercontent.com/19214100/82873333-413b3c80-9f2c-11ea-88b7-a12aa4903b19.png)

The sound for Griffin-lim is very good.
[karine.hovhanissyan_05222.wav.zip](https://github.com/espnet/espnet/files/4680592/karine.hovhanissyan_05222.wav.zip)

So thank you for your support! 
Finally I get a result on the Tacotron2 after a long time attempts. 
I'll try to find and fixTacotron's issues with text processing to avoid this excess operation (phoneme conversation) im my further  development.
I'll be so apprecieted if you advise me correct/desireble way to feed it with vocoder.
